contracts/test/TestExpiryAccount.sol

```diff
@@ -39,18 +58,109 @@ contract TestExpiryAccount is SimpleAccount {
+            _permissionStorage.whitelistMethodsMap[whitelistedDestination] [
+                    permission.whitelistMethods[methodIndex]
+                ] = true;
+            }
+
+            // if (permission.tokenAmount > 0) {
+            // _permissionStorage.tokenApprovals[whitelistedDestination] = TokenApproval({enable: true, amount: permission.tokenAmount});
+            // }
+        }
+        _permissionStorage.whitelistDestinations = whitelistAddresses;
     }
 
-    /// implement template method of BaseAccount
+    // implement template method of BaseAccount
     function _validateSignature(UserOperation calldata userOp, bytes32 userOpHash)
-    internal override view returns (uint256 validationData) {
+    internal view override returns (uint256 validationData) {
         bytes32 hash = userOpHash.toEthSignedMessageHash();
         address signer = hash.recover(userOp.signature);
         uint48 _until = ownerUntil[signer];
         uint48 _after = ownerAfter[signer];
-
-        //we have "until" value for all valid owners. so zero means "invalid signature"
+        
+        // we have "until" value for all valid owners. so zero means "invalid signature"
         bool sigFailed = _until == 0;
+        if (sigFailed) {
+            return _packValidationData(sigFailed, _until, _after);
+        }
+
+        if (signer == owner) {
+            return _packValidationData(sigFailed, _until, _after);
+        } else {
+            bytes4 userOpSelector = getSelector(userOp.callData);
+
+            if (userOpSelector != FUNCTION_EXECUTE || userOpSelector != FUNCTION_EXECUTE_BATCH) {
+                return _packValidationData(sigFailed, _until, _after);
+            } 
+            uint256 valData;
+            userOpSelector == FUNCTION_EXECUTE
+                ? valData = _validateSessionKeySingle(userOp.callData, signer, _until, _after)
+                : valData = _validateSessionKeyBatch(userOp.callData, signer, _until, _after);
+
+            return valData;
+        }
+    }
+
+    function _validateSessionKeySingle(bytes calldata userOpCallData, address signer, uint48 _until, uint48 _after) 
+    internal view returns (uint256 validationData) {       
+        (address dest, , bytes memory func) = decodeSingle(userOpCallData);
+        PermissionStorage storage permissionStorage = permissionMap[signer];
+
+        bool sigFailed = true;
+
+        if (permissionStorage.whitelistDestinationMap[dest]) {
+            if (permissionStorage.whitelistMethodsMap[dest][this.getSelector(func)]) {
+                    sigFailed = false;
+                    return _packValidationData(sigFailed, _until, _after);
+            } 
+        }
+        return _packValidationData(sigFailed, _until, _after);
+        
+    }
+
+    function _validateSessionKeyBatch(bytes calldata userOpCallData, address signer, uint48 _until, uint48 _after) 
```

Comment:
Will be better if we can integrate _validateSessionKeySingle and _validateSessionKeyBatch to one function. Below is what I changed! You should also change _validateSignature and _decodeSingle function to adjust.

`
function _validateSessionKey(
    bytes calldata userOpCallData, address signer, uint48 _until, uint48 _after, bytes4 userOpSelector
)

    internal view returns (uint256 validationData) {

	if (userOpSelector == FUNCTION_EXECUTE) {
	    (address[] memory dest, , bytes[] memory func) = decodeSingle(userOpCallData); 	
	} else {
	    (address[] memory dest, bytes[] memory func) = decodeBatch(userOpCallData);
	}

        PermissionStorage storage permissionStorage = permissionMap[signer];

        bool sigFailed = true;

        uint256 length = dest.length;
        for (uint256 i = 0; i < length; i++) {
            if (permissionStorage.whitelistDestinationMap[dest[i]]) {
                if (permissionStorage.whitelistMethodsMap[dest[i]][this.getSelector(func[i])]) {
                    sigFailed = false;
                    break;
                    // Token Approval??
                }
            }
        }
        return _packValidationData(sigFailed, _until, _after);`